modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/prepare/IgniteTypeCoercion.java

```diff
@@ -55,6 +58,50 @@ public IgniteTypeCoercion(RelDataTypeFactory typeFactory, SqlValidator validator
         super(typeFactory, validator);
     }
 
+    /** {@inheritDoc} **/
+    @Override public boolean binaryComparisonCoercion(SqlCallBinding binding) {
+        // Although it is not reflected in the docs, this method is also invoked for MAX, MIN (and other similar operators)
+        // by ComparableOperandTypeChecker. When that is the case, fallback to default rules.
+        SqlCall call = binding.getCall();
+
+        if (binding.getOperandCount() != 2 || !SqlKind.BINARY_COMPARISON.contains(call.getKind()))
+            return super.binaryComparisonCoercion(binding);
+
+        SqlValidatorScope scope = binding.getScope();
+
+        RelDataType leftType = validator.deriveType(scope, call.operand(0));
+        RelDataType rightType = validator.deriveType(scope, call.operand(1));
+
+        if (leftType.equals(rightType))
+            return super.binaryComparisonCoercion(binding);
+        else {
+            // Find the least restrictive type among the operand types
+            // and coerce the operands to that type if such type exists.
+            //
+            // An example of a least restrictive type from the javadoc for RelDataTypeFactory::leastRestrictive:
+            // leastRestrictive(INT, NUMERIC(3, 2)) could be NUMERIC(12, 2)
+            //
+            // A least restrictive type between types of different type families does not exist -
+            // the method returns null (See SqlTypeFactoryImpl::leastRestrictive).
+            //
+            RelDataType targetType = factory.leastRestrictive(Arrays.asList(leftType, rightType));
+
+            if (targetType == null || targetType.getFamily() == SqlTypeFamily.ANY)
+                return super.binaryComparisonCoercion(binding);
+            else {
+                boolean coerced = false;
+
+                if (!leftType.equals(targetType))
+                    coerced = coerceOperandType(scope, call, 0, targetType);
+
+                if (!rightType.equals(targetType))
+                    coerced &= coerceOperandType(scope, call, 1, targetType);
```

Comment:
Why "&"? Looks like we should return true if any of operands is coerced ("or" condition)
Also looks like `coerceOperandsType` can be used instead of manual operands check.

Reply:
Good question. Actually, Calcite doesn't take in account the result of this call:

```
if (b) {
      // Coerce type first.
      if (callBinding.isTypeCoercionEnabled()) {
        TypeCoercion typeCoercion = callBinding.getValidator().getTypeCoercion();
        // For comparison operators, i.e. >, <, =, >=, <=.
        typeCoercion.binaryComparisonCoercion(callBinding);
      }
b = super.checkOperandTypes(callBinding, false);
```

'OR' works too. I checked. Fixed.